source/simulator/stokes_matrix_free.cc

```diff
@@ -2285,7 +2285,10 @@ namespace aspect
                 }
 
             diagonal_matrix.compress(VectorOperation::add);
-            mg_matrices_A_block[level].set_diagonal(diagonal_matrix.get_vector());
+
+            dealii::LinearAlgebra::distributed::Vector<gmg_number> diagonal;
+            diagonal = diagonal_matrix.get_vector();
+            mg_matrices_A_block[level].set_diagonal(diagonal);
```

Comment:
This is a bit ugly, but for reasons of instantiations inside the dealii ConstraintMatrix classes, we must compute the diagonal of the GMG matrices (for curved boundary, tangential constraints) in double precision. Thus, we need to create a single precision diagonal vector and copy the contents after computing the diagonal and before setting them in the MatrixFree operators. This should not cause a significant increase in runtime.

Reply:
please leave a comment before the assignment that this does the double->float conversion.